diff --git a/cups/auth.c b/cups/auth.c
index de4dc62..14ddc09 100644
--- a/cups/auth.c
+++ b/cups/auth.c
@@ -393,6 +393,13 @@ cupsDoAuthentication(
     httpSetAuthString(http, "Digest", digest);
   }
 
+  if (http->status == HTTP_UNAUTHORIZED && http->digest_tries >= 3)
+  {
+    DEBUG_printf(("1cupsDoAuthentication: too many authentication tries (%d)", http->digest_tries));
+    http->status = HTTP_AUTHORIZATION_CANCELED;
+    return(-1);
+  }
+
   DEBUG_printf(("1cupsDoAuthentication: authstring=\"%s\"", http->authstring));
 
   return (0);
